core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/NotifiableRegistry.java

```diff
@@ -0,0 +1,97 @@
+import org.optaplanner.core.impl.domain.variable.descriptor.VariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.listener.SourcedVariableListener;
+import org.optaplanner.core.impl.score.director.InnerScoreDirector;
+
+final class NotifiableRegistry<Solution_> {
+
+    private final InnerScoreDirector<Solution_, ?> scoreDirector;
+    private final List<VariableListenerNotifiable<Solution_>> notifiableList = new ArrayList<>();
+    private final Map<EntityDescriptor<?>, Set<VariableListenerNotifiable<Solution_>>> sourceEntityToNotifiableMap =
+            new LinkedHashMap<>();
+    private final Map<VariableDescriptor<?>, List<VariableListenerNotifiable<Solution_>>> sourceVariableToNotifiableMap =
+            new LinkedHashMap<>();
+    private int nextGlobalOrder = 0;
+
+    NotifiableRegistry(InnerScoreDirector<Solution_, ?> scoreDirector) {
+        this.scoreDirector = scoreDirector;
+        for (EntityDescriptor<Solution_> entityDescriptor : scoreDirector.getSolutionDescriptor().getEntityDescriptors()) {
+            sourceEntityToNotifiableMap.put(entityDescriptor, new LinkedHashSet<>());
+            for (VariableDescriptor<Solution_> variableDescriptor : entityDescriptor.getDeclaredVariableDescriptors()) {
+                sourceVariableToNotifiableMap.put(variableDescriptor, new ArrayList<>());
+            }
+        }
+    }
+
+    void registerShadowVariableListener(ShadowVariableDescriptor<Solution_> shadowVariableDescriptor,
+            VariableListener<Solution_, ?> variableListener) {
+        int globalOrder = shadowVariableDescriptor.getGlobalShadowOrder();
+        VariableListenerNotifiable<Solution_> notifiable =
+                new VariableListenerNotifiable<>(scoreDirector, variableListener, globalOrder);
+        for (VariableDescriptor<Solution_> source : shadowVariableDescriptor.getSourceVariableDescriptorList()) {
+            registerNotifiable(source, notifiable);
+        }
+        notifiableList.add(notifiable);
+        nextGlobalOrder = Math.max(nextGlobalOrder, globalOrder + 1);
+    }
+
+    void registerSourcedVariableListener(SourcedVariableListener<Solution_, ?> variableListener) {
+        VariableListenerNotifiable<Solution_> notifiable =
+                new VariableListenerNotifiable<>(scoreDirector, variableListener, nextGlobalOrder);
+        VariableDescriptor<Solution_> source = variableListener.getSourceVariableDescriptor();
+        registerNotifiable(source, notifiable);
+        notifiableList.add(notifiable);
+        nextGlobalOrder++;
+    }
+
+    private void registerNotifiable(VariableDescriptor<?> source, VariableListenerNotifiable<Solution_> notifiable) {
+        sourceVariableToNotifiableMap.get(source).add(notifiable);
+        sourceEntityToNotifiableMap.get(source.getEntityDescriptor()).add(notifiable);
+    }
+
+    void sort() {
+        Collections.sort(notifiableList);
+    }
```

Comment:
I don't like the fact internal consistency of the implentation needs to be maintained from the outside.